src/infra/ingest-datafusion/src/visitor.rs

```diff
@@ -0,0 +1,254 @@
+        if self.maybe_source_event.is_none() {
+            if self.maybe_source_name.is_none()
+                || self.maybe_source_name == Some(e.source_name.as_str())
+            {
+                self.maybe_source_event = Some(e.clone().into());
+            }
+        } else {
+            // Encountered another source - if `source_name` was not specified we
+            // return ambiguity error
+            if self.maybe_source_name.is_none() {
+                return Err(SourceNotFoundError::new(
+                    None::<&str>,
+                    "Explicit source name is required to pick between several active push sources",
+                )
+                .into());
+            }
+        }
+
+        Ok(())
+    }
+
+    fn handle_set_vocab(&mut self, e: &SetVocab) {
+        if self.maybe_set_vocab.is_none() {
+            self.maybe_set_vocab = Some(e.clone());
+        }
+    }
+
+    fn handle_seed(&mut self, e: &Seed) {
+        assert_eq!(e.dataset_kind, DatasetKind::Root);
+    }
+}
+
+impl<'a> MetadataChainVisitor for DataWriterDataFusionMetaDataStateVisitor<'a> {
+    type VisitError = ScanMetadataError;
+
+    fn visit(&mut self, (_, block): HashedMetadataBlockRef) -> Result<Decision, Self::VisitError> {
+        match &block.event {
+            MetadataEvent::SetDataSchema(e) => {
+                self.handle_set_data_schema(e)?;
+
+                self.next_block_flags -= Flag::SET_DATA_SCHEMA;
+            }
+            MetadataEvent::AddData(e) => {
+                self.handle_add_data(e);
+            }
+            MetadataEvent::SetPollingSource(e) => {
+                self.handle_set_polling_source(e)?;
+
+                if self.maybe_source_name.is_none() {
+                    self.next_block_flags -= Flag::SET_POLLING_SOURCE;
+                }
+            }
+            MetadataEvent::AddPushSource(e) => {
+                self.handle_add_push_source(e)?;
+
+                if self.maybe_source_name.is_some() && self.maybe_source_event.is_some() {
+                    self.next_block_flags -= Flag::ADD_PUSH_SOURCE;
+                }
+            }
+            MetadataEvent::SetVocab(e) => {
+                self.handle_set_vocab(e);
+
+                self.next_block_flags -= Flag::SET_VOCAB;
+            }
+            MetadataEvent::Seed(e) => {
+                self.handle_seed(e);
+
+                self.next_block_flags -= Flag::SEED;
+            }
+            MetadataEvent::DisablePollingSource(_) | MetadataEvent::DisablePushSource(_) => {
+                unimplemented!("Disabling sources is not yet fully supported")
+            }
+            MetadataEvent::ExecuteTransform(_)
+            | MetadataEvent::SetTransform(_)
+            | MetadataEvent::SetAttachments(_)
+            | MetadataEvent::SetInfo(_)
+            | MetadataEvent::SetLicense(_) => {
+                unreachable!()
+            }
+        }
+
+        Ok(Decision::NextOfType(self.next_block_flags))
```

Comment:
Will the `NextOfType` with no bits set be automatically treated as `Stop`?

Reply:
btw, no -- fixed it to an explicit stop. thank you for this find!